### PR TITLE
refactor(skill-forge): structural cleanups from code quality audit

### PR DESCRIPTION
**Pull request by j5ik2o**
## 概要

skill-forge 監査（#69 の続き）で報告した構造的問題4件の修正です。挙動を変えるのは重複クエリとdelta方向の2点のみで、他は等価な構造整理です。

### 1. frontmatterパーサーの3重実装を統一（`43e1b34`）

`utils.parse_skill_md`（手書き行パーサー）、`quick_validate`、`generate_openai_yaml`（regex+yamlのコピー2つ）がそれぞれ独自にfrontmatterを解釈していました。`utils.parse_frontmatter` に一本化し、ブロックスカラー・クォート・エラー文言のセマンティクスを1箇所に集約しました。

- 共有ヘルパーのimportに伴い、`quick_validate` / `generate_openai_yaml` の起動形式をモジュール形式（`python -m scripts.quick_validate`）に変更（`run_loop`/`package_skill` と同形式に統一）。SKILL.md・`references/openai_yaml.md`・`verify.sh` を更新し、`agents/openai.yaml` を再生成（interface値は既存のものを維持）
- 手書きパーサーと異なり、不正なYAML frontmatterは明示的に `ValueError` になります（`quick_validate` は従来からYAML妥当性を要求していたため、実質的な互換性影響はなし）

### 2. eval結果を位置キーに変更し、重複クエリの無音マージを解消（`14efdde`）

`run_eval` がクエリ文字列で結果をマージしていたため、eval set内の重複クエリが1エントリに潰れ、`run_loop` のクエリ文字列によるtrain/test再分割でtest結果がtrainに漏れる可能性がありました。結果を入力順の位置キーで管理し、`run_loop` は長さでスライスして分割するようにしました。

### 3. run_loop履歴のlegacyキーを削除（`14efdde`）

history要素が `train_*` と legacy `passed/failed/total/results` の二重形状を持ち、`improve_description`・`generate_report` にフォールバック読み取りが散在していました。書き手・読み手とも `train_*` 形状に一本化しました（`improve_description` CLIの履歴出力形状も統一）。

### 4. benchmark deltaの方向を明示化（`efed664`）

delta が「最初の2つのconfig」＝ディレクトリのアルファベット順に依存し、`baseline`/`skill` のような命名では符号が黙って反転していました。既知のbaseline名（`without_skill`/`old_skill`/`baseline`）をbaseline位置に固定し、`delta.comparison`（例: `"with_skill - without_skill"`）で方向を自己記述化。`benchmark.md` にも表示します。

## テスト

- `bash scripts/ci/verify.sh`: ユニットテスト 125 件（+6件追加）パス + スキルバリデーション（auto / codex strict）通過
- 追加テスト: 不正YAML・非dict frontmatter / 重複クエリの独立性と順序 / holdout分割のスライス整合とlegacyキー不在 / delta方向（`baseline`/`skill` と `with_skill`/`without_skill`）

## 関連

- 監査由来の修正第1弾: #69

---

> [!NOTE]
> **Medium Risk**
> Changes trigger-eval aggregation and benchmark delta semantics (user-visible for skill optimization), but scope is confined to skill-forge tooling with added regression tests.
> 
> **Overview**
> **Structural cleanup** in skill-forge after a code-quality audit: shared **`parse_frontmatter`** in `utils` replaces three separate SKILL.md frontmatter parsers; **`quick_validate`** and **`generate_openai_yaml`** now use it and are invoked as **`python -m scripts.*`** (docs, `verify.sh`, and `openai.yaml` source hash updated).
> 
> **Behavior fixes:** **`run_eval`** keeps one result per eval-set **index** (duplicate query text no longer merges); **`run_loop`** splits train/test by **slicing** ordered results, and history drops legacy **`passed`/`results`** keys in favor of **`train_*` only**. **`aggregate_benchmark`** pins known baseline config names (`without_skill`, `old_skill`, `baseline`) so delta sign is stable and adds **`delta.comparison`** plus a **Delta** line in **`benchmark.md`**. **`generate_report`** maps HTML columns by **position**, not query string.
> 
> New unit tests cover delta orientation, duplicate queries, holdout slicing, frontmatter edge cases, and report HTML.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6f772e5861fdc205568897275b3db51d534eb92c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: plugins/agent-skills/skills/skill-forge/scripts/aggregate_benchmark.py [hunk 1]
```diff
@@ -235,7 +254,9 @@ def aggregate_results(results: dict) -> dict:
     run_summary["delta"] = {
         "pass_rate": f"{delta_pass_rate:+.2f}",
         "time_seconds": f"{delta_time:+.1f}",
-        "tokens": f"{delta_tokens:+.0f}"
+        "tokens": f"{delta_tokens:+.0f}",
+        # Self-describing orientation so the sign never depends on key order
+        "comparison": f"{primary_name} - {baseline_name}" if baseline_name else (primary_name or ""),
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply plugins/agent-skills/skills/skill-forge/scripts/aggregate_benchmark.py:264]: Valid finding — fixed in 1fefc4b. `order_configs` now places the recognized baseline in the second slot (the position the delta actually compares against), so benchmarks with more than two configurations still compute primary minus baseline. Added a regression test with three configs (`alt`/`baseline`/`skill`) asserting `comparison == "alt - baseline"`.

File: plugins/agent-skills/skills/skill-forge/scripts/generate_report.py [hunk 2]
```diff
@@ -206,17 +206,17 @@ def generate_html(data: dict, auto_refresh: bool = False, skill_name: str = "")
...
         iteration = h.get("iteration", "?")
-        train_passed = h.get("train_passed", h.get("passed", 0))
-        train_total = h.get("train_total", h.get("total", 0))
+        train_passed = h.get("train_passed", 0)
+        train_total = h.get("train_total", 0)
         test_passed = h.get("test_passed")
         test_total = h.get("test_total")
         description = h.get("description", "")
-        train_results = h.get("train_results", h.get("results", []))
+        train_results = h.get("train_results", [])
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply plugins/agent-skills/skills/skill-forge/scripts/generate_report.py:219]: Valid finding — fixed in 1fefc4b. Report cells are now resolved by position instead of a query-text dict, matching the stable eval-set order that run_eval guarantees, so duplicate query columns render their own outcomes. Also guarded `test_results: None` which previously crashed `aggregate_runs`. Added tests in tests/test_generate_report.py.

File: plugins/agent-skills/skills/skill-forge/scripts/aggregate_benchmark.py [hunk 3]
```diff
@@ -220,12 +236,15 @@ def aggregate_results(results: dict) -> dict:
...
         }
 
-    # Calculate delta between the first two configs (if two exist)
+    # Calculate delta between the primary and baseline configs (if two exist)
     if len(configs) >= 2:
-        primary = run_summary.get(configs[0], {})
-        baseline = run_summary.get(configs[1], {})
+        primary_name, baseline_name = configs[0], configs[1]
+        primary = run_summary.get(primary_name, {})
+        baseline = run_summary.get(baseline_name, {})
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply plugins/agent-skills/skills/skill-forge/scripts/aggregate_benchmark.py:248]: ご指摘の通りですが、この問題は同内容のBugbot指摘を受けて 1fefc4b で修正済みです（本コメントはefed664時点のコードに対するもの）。`order_configs()` は認識済みベースラインを末尾ではなく2番目のスロット（deltaの比較位置）に配置するようになり、3構成（`alt`/`baseline`/`skill`）で `comparison == "alt - baseline"` を検証する回帰テストも追加しています。

File: plugins/agent-skills/skills/skill-forge/scripts/generate_report.py [hunk 4]
```diff
@@ -23,7 +23,7 @@ def generate_html(data: dict, auto_refresh: bool = False, skill_name: str = "")
     train_queries: list[dict] = []
     test_queries: list[dict] = []
     if history:
-        for r in history[0].get("train_results", history[0].get("results", [])):
+        for r in history[0].get("train_results", []):
             train_queries.append({"query": r["query"], "should_trigger": r.get("should_trigger", True)})
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply plugins/agent-skills/skills/skill-forge/scripts/generate_report.py:27]: ご指摘の通りですが、この問題も同内容のBugbot指摘を受けて 1fefc4b で修正済みです（本コメントはefed664時点のコードに対するもの）。`train_by_query`/`test_by_query` のdictは削除し、run_evalが保証する入力順序を利用した位置ベースのセル解決に変更しました。重複クエリ列が各自の結果を表示することは tests/test_generate_report.py で検証しています。

File: plugins/agent-skills/skills/skill-forge/scripts/utils.py [hunk 5]
```diff
@@ -1,14 +1,19 @@
...
 import sys
 from contextlib import contextmanager
 from pathlib import Path
 
+import yaml
+
 CLI_CLAUDE = "claude"
 CLI_CODEX = "codex"
 
+FRONTMATTER_PATTERN = re.compile(r"^---\n(.*?)\n---", re.DOTALL)
```
[the remark this answers is not in the document]
Comment (j5ik2o) [reply plugins/agent-skills/skills/skill-forge/scripts/utils.py:15]: 妥当な指摘です。6f772e5 で修正しました。`parse_frontmatter` が正規表現マッチ前に CRLF を LF に正規化するようにし、CRLF の SKILL.md を `write_bytes` で作成する回帰テストを追加しています。
